cogeo_mosaic/backends/sqlite.py

```diff
@@ -0,0 +1,237 @@
+    MosaicError,
+    MosaicExistsError,
+    MosaicNotFoundError,
+)
+from cogeo_mosaic.logger import logger
+from cogeo_mosaic.mosaic import MosaicJSON
+from cogeo_mosaic.utils import bbox_union
+
+try:
+    import sqlite3
+    from sqlite3 import Connection, OperationalError
+
+    sqlite3.register_adapter(dict, json.dumps)
+    sqlite3.register_converter("JSON", json.loads)
+except ImportError:  # pragma: nocover
+    sqlite3 = None  # type: ignore
+    Connection = None  # type: ignore
+    OperationalError = None  # type: ignore
+
+
+@attr.s
+class SQLiteBackend(BaseBackend):
+    """SQLite Backend Adapter."""
+
+    db_name: str = attr.ib(init=False)
+    mosaic_name: str = attr.ib(init=False)
+
+    db: Connection = attr.ib(init=False)
+
+    _backend_name = "SQLite"
+    _metadata_quadkey: str = "-1"
+
+    def __attrs_post_init__(self):
+        """Post Init: parse path connect to Table.
+
+        A path looks like
+
+        sqlite:///{db_name}:{mosaic_name}
+
+        """
+        assert sqlite3 is not None, "'sqlite3' must be installed to use SQLiteBackend"
+
+        parsed = urlparse(self.path)
+        mosaic_info = parsed.path.lstrip("/").split(":")
+        self.db_name = mosaic_info[0]
+        self.mosaic_name = mosaic_info[1]
+
+        self.db = sqlite3.connect(self.db_name, detect_types=sqlite3.PARSE_DECLTYPES)
+        self.db.row_factory = sqlite3.Row
+
+        logger.debug(f"Using SQLite backend: {self.db_name}")
+        super().__attrs_post_init__()
+
+    def close(self):
+        """Close SQLite connection."""
+        self.db.close()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers."""
+        self.close()
+
+    @property
+    def _quadkeys(self) -> List[str]:
+        """Return the list of quadkey tiles."""
+        with self.db:
+            rows = self.db.execute(
+                f"SELECT quadkey FROM {self.mosaic_name} WHERE quadkey != ?;",
+                (self._metadata_quadkey,),
+            ).fetchall()
+
+        return [r["quadkey"] for r in rows]
+
+    def write(self, overwrite: bool = False):
+        """Write mosaicjson document to an SQLite database.
+
+        Args:
+            overwrite (bool): delete old mosaic items inthe Table.
+
+        Returns:
+            dict: dictionary with metadata constructed from the sceneid.
+
+        Raises:
+            MosaicExistsError: If mosaic already exists in the Table.
+
+        """
+        if self._mosaic_exists():
+            if not overwrite:
+                raise MosaicExistsError(
+                    f"'{self.mosaic_name}' Table already exists in {self.db_name}, use `overwrite=True`."
+                )
+            self.delete()
+
+        logger.debug(f"Creating '{self.mosaic_name}' Table in {self.db_name}.")
+        with self.db:
+            self.db.execute(
+                f"CREATE TABLE {self.mosaic_name} (quadkey TEXT NOT NULL, value JSON NOT NULL);",
+            )
```

Comment:
I just took https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager as example 🤷‍♂️ 
I don't think it will close the connection.

Reply:
I didn't know that

> Connection object used as context manager only commits or rollbacks transactions, so the connection object should be closed manually

That's why you can get by without ever doing `db.commit()`. Note that I think the commit is expensive, so you can get better performance for bulk ingest if you do your writes in chunks or all at once with a commit at the end.

Reply:
>  Note that I think the commit is expensive, so you can get better performance for bulk ingest if you do your writes in chunks or all at once with a commit at the end.

I'm not sure to get this!

Reply:
I think this

```py
for item in items:
	db.insert()
	db.commit()
```

is probably much slower than

```py
for item in items:
	db.insert()

db.commit()
```

and likewise this

```py
for item in items:
	with db:
		db.insert()
```

does a commit on each iteration, and would probably be slower than

```py
with db:
	for item in items:
		db.insert()
```

Reply:
oh this is only for `update` part, not in general.

Reply:
Well in this case most methods only do one command at a time. It's relevant mostly for bulk transactions

Reply:
I'm not sure I get this fully.